WIP: Element screenshots

mattzeunert opened the pull request:
**Summary**

Very WIP 🏗

- take full-page screenshot (created a [separate PR for this](https://github.com/GoogleChrome/lighthouse/pull/9064))
- capture element bounding boxes
- render element screenshot in report

Todo:
- [ ] figure out final UX
- [ ] we should scroll down the page once before capturing element positions or the screenshot
   - otherwise page content size may change on scroll, e.g. because of lazy loading images/ads
   - example: footer promo slider at the bottom of the BBC homepage
   - Ticket: https://github.com/GoogleChrome/lighthouse/issues/8647
otherwise page content size may change on scroll (e.g. on the BBC homepage) 
- [x] sort out how chrome devtools instantiates detailsrenderer
    - I think this may actually be resolved now, seems like [Chrome no longer has custom code to initialize the details renderer](https://cs.chromium.org/chromium/src/third_party/blink/renderer/devtools/front_end/audits/lighthouse/report.js?l=3853&rcl=df13cdcb4cd91c79eb63bdd33267b8e67f9f43b3)
- [ ] don't use clippath (see Patrick's note on browser support, also there's no reason we need it)
- [ ] do we need to check if the element is not visible and then not capture the bounding box so that we don't show a screenshot that doesn't show the element (e.g. because it's overflowing a scroll container or there's an overlay on top of it)
- [ ] handle elements that are outside the captured screenshot area (e.g. for pages > 16000px)

Known issues:
- if resizing the window changes the layout then the element marker positions will be off (e.g. this happens for position fixed elements, or if the page uses `vh`)
- elements in `overflow: scroll` containers may be hidden
- elements can be obscured by full-screen overlays

**Related Issues/PRs**

fixes #6683 
fixes #7327

mattzeunert commented on lighthouse-core/report/html/renderer/details-renderer.js:
```diff
@@ -355,15 +355,43 @@ class DetailsRenderer {
+    snippetEl.innerText = item.snippet;
+    element.appendChild(snippetEl);
+
+    const fullpageScreenshotUrl = __LIGHTHOUSE_JSON__.audits['full-page-screenshot'].details.data;
```
What's a good way to access this in the renderer?

patrickhulce replied:
I would say it should probably be passed into a `DetailsRenderer` as the context of the report, almost like `setTemplateContext`

mattzeunert replied:
Passing it into the `DetailsRenderer` constructor now.

patrickhulce commented on lighthouse-core/gather/gatherers/full-page-screenshot.js:
```diff
@@ -0,0 +1,45 @@
+    height = Math.min(maxScreenshotHeight, height);
+
+    await driver.sendCommand('Emulation.setDeviceMetricsOverride', {
+      // todo: figure out what this means and set appropriately
```
😆 I wonder how many devtools protocol settings we should add this comment to 🤔

patrickhulce commented on lighthouse-core/report/html/renderer/element-screenshot-renderer.js:
```diff
@@ -0,0 +1,97 @@
+    const tmpl = dom.cloneTemplate('#tmpl-lh-element-screenshot', templateContext);
+    const previewContainer = dom.find('.lh-element-screenshot', tmpl);
+
+    // todo: support desktop screenshots
```
ideally just passed around along with the `fullPageScreenshot` artifact so there's no support for specific screen sizes :)

mattzeunert replied:
Added support for desktop screenshots now, i.e. screenshots that are wider than the element screenshot display area.

Right now if there's a big element highlighted on the screenshot it'll just be cut off with `overflow: hidden`. Maybe in the future we could address that and maybe zoom out a bit? Probably not something that needs to be part of v1.

patrickhulce replied:
Gotcha yeah let's punt on that for v1 :)

patrickhulce commented on types/artifacts.d.ts:
```diff
@@ -127,6 +127,8 @@ declare global {
       /** Information about tap targets including their position and size. */
       TapTargets: Artifacts.TapTarget[];
+      /** Screenshot of the entire page (rather than just the above the fold content). */
+      FullPageScreenshot: {data: string}
```
dimensions here would be 👌

patrickhulce commented on lighthouse-core/report/html/report-styles.css:
```diff
@@ -1366,3 +1366,29 @@
+.lh-element-screenshot__content {
+  position: absolute;
+  right: 0px;
+  top: -190px;
```
this is based on the dynamic height we set it js? anyway to fix it or not have the hanging magic value? :)

mattzeunert replied:
It's so that the screenshot doesn't cover the table item you're hovering over and instead appears above it. Not really based on anything though.

Do you think showing the screenshot above the table item on hover is the way to go? And then on mobile we could expand the table item to include the screenshot when the user taps on it.

patrickhulce commented on lighthouse-core/report/html/renderer/details-renderer.js:
```diff
@@ -352,15 +352,43 @@ class DetailsRenderer {
+    if (item.boundingRect) {
+      /** @type {Element} */
+      let elementScreenshot;
+      element.addEventListener('mouseenter', () => {
```
not sure what our stance is on JS in report now

@paulirish 🤓 🔫 , CSS-only impl? ;)

mattzeunert replied:
I can't find the comment, but I think we decided that's ok on the snippet renderer PR.

Not sure how feasible this would be to do without JS, browsers might not like having a hundred 500kb base64 images on one page.

patrickhulce replied:
Oh that sounds right. I was remembering some I/O convo about going back and forth to make things easier on PSI, but we did already break that barrier.

patrickhulce commented on lighthouse-core/report/html/renderer/element-screenshot-renderer.js:
```diff
@@ -0,0 +1,97 @@
+    const clipId = 'clip-' + Math.floor(Math.random() * 100000000);
+    mask.style.width = previewWidth + 'px';
+    mask.style.height = previewHeight + 'px';
+    mask.style.clipPath = 'url(#' + clipId + ')';
```
clever! I think the chart is scarier than reality given this narrow usecase but still might be troublesome for us? https://caniuse.com/#search=clipPath

would you mind explaining in words what the approach is here? I think I'm getting it but not sure why the mask is necessary

mattzeunert replied:
There isn't anything that can't be done with divs positioned over the image, will change it.

mattzeunert commented:
> do you think you could update the artifacts so we can play around with it in the now.sh deployment report on the PR

Done now 👍

mattzeunert commented on lighthouse-core/audits/accessibility/axe-audit.js:
```diff
@@ -53,6 +53,8 @@ class AxeAudit extends Audit {
           selector: Array.isArray(node.target) ? node.target.join(' ') : '',
           path: node.path,
           snippet: node.html || node.snippet,
+          textContent: node.textContent,
```
I added the text content to the rendered elements to make them easier to read. If we only show the element screenshots on hover/tap then I think it's worth showing. What do you think?

Do we need the property for that? Yes because the HTML may be truncated, so we can't extract the text from it. But we can limit it to 80 characters or so.

![Screenshot 2019-05-15 at 09 48 31](https://user-images.githubusercontent.com/1303660/57762067-ab54bb00-76f6-11e9-8e81-64f79db8d0d7.png)

mattzeunert replied:
Maybe we should call it `text` and we can also count `alt` or `aria-label` attributes if there's no textContent

patrickhulce replied:
Hm, might be difficult if we're specifically flagging elements that are using `alt`/`aria-label` incorrectly 😆 

how about we stick to textContent for now but limit to 80 characters as you suggest? 👍

mattzeunert replied:
I think using `alt/aria-label` is quite helpful, e.g. for buttons that just consist of an image.

Created a PR here: https://github.com/GoogleChrome/lighthouse/pull/8961

patrickhulce replied:
Oh if you've already done it then more power too ya! 👍 ❤️

patrickhulce commented:
Nice thanks for deploying it! The experience is A-MAZ-ING 😍 


Few findings:

- Small screen sizes it gets a little weird that it goes full screen and completely covers the node, maybe we scale it down a bit so we can shift it to the side?
![image](https://user-images.githubusercontent.com/2301202/57786047-6df12d00-76f8-11e9-8cf5-93e35202b708.png)

- This is the sort of thing I was worried about with textContent, maybe we actually want innerText or perhaps the 80 character limit is enough?
![image](https://user-images.githubusercontent.com/2301202/57786174-a6910680-76f8-11e9-97be-0cc371f4b3b0.png)

mattzeunert commented:
> This is the sort of thing I was worried about with textContent, maybe we actually want innerText or perhaps the 80 character limit is enough?

Addressed in [this PR](https://github.com/GoogleChrome/lighthouse/pull/8961).

> It wasn't able to get the screenshot for these two links here, not sure whats going on with that

They have a visible size of 0x0px – do you think we should leave some kind of indication if there's no screenshot?

> Sometimes the full-size preview isn't really what we want I think, we might actually prefer context for where it is on the page? Example where we have a ton of duplicate lighthouse images on the page so I still don't know which one this is talking about.

Maybe if the element is taller than 60px we try to show where on the page it is and zoom out by 2x or 4x, and for smaller ones we do something similar to what we have now?

> Small screen sizes it gets a little weird that it goes full screen and completely covers the node, maybe we scale it down a bit so we can shift it to the side?

I changed the CSS so that the hover always appears above the table item. Let's see how the zoomed out stuff from above works and see if it can help with this too.

What do you mean by "shift it to the side"?

> Another small screen size failure case where the important part of the screenshot is outside the window.

Will do this:
- if there's a hover contain it within the window
- if the page is narrower than say 800px users have to tap the line with the element

mattzeunert commented:
@patrickhulce I just introduced zooming out for large elements, and it's a big improvement!

Played around with always zooming out, but for e.g. small tap targets the immediate environment is more important than the page context.

The screenshot positioning feels a bit awkward to me sometimes, but not sure if there's a better approach. Maybe we could always put them in a corner of the screen?

Also, here are two screenshots from a desktop report:



![Screenshot 2019-05-16 at 16 55 37](https://user-images.githubusercontent.com/1303660/57868754-7da56a00-77fb-11e9-902c-41b293578950.png)
![Screenshot 2019-05-16 at 16 55 41](https://user-images.githubusercontent.com/1303660/57868755-7da56a00-77fb-11e9-8132-f071c189e67c.png)

patrickhulce commented:
Woah yes zooming out makes a massive difference, great work!!

> do you think we should leave some kind of indication if there's no screenshot?

Yeah I think that'd be a good call 👍 

> What do you mean by "shift it to the side"?

I just meant always make sure we can see a portion of the table item, i.e. let's move the screenshot out of the way a little bit.

> Maybe we could always put them in a corner of the screen?

This is an interesting idea, I like it! A consistent place for the screenshots to appear might be nice.

mattzeunert commented:
> > Maybe we could always put them in a corner of the screen?
> 
> This is an interesting idea, I like it! A consistent place for the screenshots to appear might be nice.
 
Tried that now, and it's really nice on a big screen (except for the flicker when hovering over the td padding).
 
I was originally thinking we'd put the screenshot inside the td on mobile, but actually putting it outside seems better because we don't have to shrink it down, e.g. in the 3-column tap targets table.
 
On small and medium size screens the screenshot ends up overlapping the main content. Maybe we should show the image at the top of the screen if it overlaps? I thought about trying it out but then realized how much of a pain it is to get the location of the `td` in the screenshot renderer 🙂

---

Do you think we should try to re-use the `final-screenshot` instead of duplicating the top of the page in `full-page-screenshot`?

- pro: we'd save some JSON size
- pro: part of the full-page screenshot would be higher quality
- con: it adds complexity
- con: visible change in jpeg quality in element screenshots

My feeling is "no".

I assume `full-page-screenshot` can't be used as `final-screenshot` since its quality is too low and it's too big.

patrickhulce commented:
> My feeling is "no".

👍

Maybe in the future we can explore that but separate for now seems fine to me.

mattzeunert commented:
@patrickhulce Been thinking of creating a PR for the full-page screenshot artifact/audit, since I think we know what we want there. Any reason we'd want to hold off on that, or do you think I should go ahead?

patrickhulce commented:
Sounds like a great place to start to me @mattzeunert! The only hesitation I can see would be if we don't want the immediate byte impact in the LHR if nothing is using it yet, but maybe we can leave it ungrouped or something?

connorjclark commented:
hey @mattzeunert we're wanting to land this (mostly for #10517) and willing to invest eng time to get it the rest of the way. mind if we take it from here, or would you be able to pick this back up?

I'll go through the PR tomorrow and leave my thoughts.

mattzeunert commented:
Hey @connorjclark! Having someone on the team pick it up sounds good 👍

connorjclark commented:
I forked and started work here: https://github.com/GoogleChrome/lighthouse/compare/elscreens
